components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java

```diff
@@ -0,0 +1,110 @@
+
+/**
+ * Update the Identity Provider application role configurations in Service Provider model from the API model.
+ */
+public class UpdateIdpAppRoleConfigurations implements UpdateFunction<ServiceProvider, List<IdpAppRoleConfig>> {
+
+    @Override
+    public void apply(ServiceProvider serviceProvider, List<IdpAppRoleConfig> idpAppRoleConfigurations) {
+
+        if (idpAppRoleConfigurations != null) {
+            updateIdpAppRoleConfigurations(serviceProvider, idpAppRoleConfigurations);
+        }
+    }
+
+    /**
+     * Update the Identity Provider application role configurations in Service Provider model from the API model.
+     *
+     * @param serviceProvider          Service Provider to be updated.
+     * @param idpAppRoleConfigurations Identity Provider application role configurations from API model.
+     */
+    private void updateIdpAppRoleConfigurations(ServiceProvider serviceProvider,
+                                                List<IdpAppRoleConfig> idpAppRoleConfigurations) {
+
+        List<String> attributeStepFIdPs = getAttributeStepFIdPs(serviceProvider);
+        AppRoleMappingConfig[] appRoleMappingConfigs = getApplicationRoleMappingConfig(
+                attributeStepFIdPs, idpAppRoleConfigurations);
+        serviceProvider.setApplicationRoleMappingConfig(appRoleMappingConfigs);
+    }
+
+    /**
+     * Get the federated identity provider names in the attribute step.
+     *
+     * @param serviceProvider Service Provider to be updated.
+     * @return List of federated identity provider names in the attribute step.
+     */
+    private List<String> getAttributeStepFIdPs(ServiceProvider serviceProvider) {
+
+        AuthenticationStep attributeAuthStep = serviceProvider.getLocalAndOutBoundAuthenticationConfig().
+                getAuthenticationStepForAttributes();
+        IdentityProvider[] authStepFederatedIdentityProviders = null;
+        if (attributeAuthStep == null) {
+            attributeAuthStep = Lists.newArrayList(serviceProvider.getLocalAndOutBoundAuthenticationConfig().
+                            getAuthenticationSteps()).stream().filter(AuthenticationStep::isAttributeStep).findFirst()
+                    .orElse(null);
+        }
+        if (attributeAuthStep != null) {
+            authStepFederatedIdentityProviders = attributeAuthStep.getFederatedIdentityProviders();
+        }
+        if (authStepFederatedIdentityProviders != null) {
+            return Arrays.stream(authStepFederatedIdentityProviders).
+                    map(IdentityProvider::getIdentityProviderName).collect(Collectors.toList());
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Get the application role mapping configurations for the service provider to be updated.
+     *
+     * @param attributeStepFIdPs List of federated identity provider names in the attribute step.
+     * @param idpAppRoleConfigs  List of IdpAppRoleConfig
+     * @return AppRoleMappingConfig[]
```

Comment:
complete the comment

Reply:
Addressed in 7cafe3dda2f22455f4845ec8dae1e99892aa824a